PWGHF: Implementation of Lc --> K0s+p

chiarazampolli opened the pull request:
some changes

update

Expect that v0s are already built and preselected

some more work

A bit further, but not compiling

just a temp commit

Compiling version - still refinements needed

further work

temp

not working

not working again, needed for debug

ppp

working version - cuts still to be tuned

chiarazampolli committed:
First version of skimming for Lc

some changes

update

Expect that v0s are already built and preselected

some more work

A bit further, but not compiling

just a temp commit

Compiling version - still refinements needed

further work

temp

not working

not working again, needed for debug

ppp

working version - cuts still to be tuned

clang

few changes

clang

further work

forgotten file

forgotten file

Fixing copy/paste error

clang

chiarazampolli commented:
@vkucera , wait before reviewing, there are some things I want to improve. 
Thanks!

chiarazampolli commented:
Hello @vkucera , @ginnocen ,
Debug plots are still missing, but I think that the review of this can start.
Thanks!
Chiara

chiarazampolli committed:
Selector for Lc --> K0s+p added

clang-format

chiarazampolli commented:
Hello @vkucera , @ginnocen , @nzardosh ,

I pushed the first version of the selector. Note that:
- the cuts for the selection are taken from https://alice-notes.web.cern.ch/system/files/notes/analysis/844/2020-11-10-ALICE_analysis_note_0.pdf
- the bayes PID used in the AN note is not yet there, as I think it is not in the framework yet (@njacazio )
- The code runs, but I did not check what gets selected yet.

Since we agreed at the meeting on Tue (9.02) that the review for this PR can start, I kept this commit separate. Let me know if you want that I squash them.

Next is the task to plot the observables of the selected candidates (equivalent to taskD0.cxx).

chiarazampolli committed:
First version of histogramming task - not compilinggit add Analysis/Tasks/PWGHF/CMakeLists.txt Analysis/Tasks/PWGHF/HFCandidateCreatorCascade.cxx Analysis/Tasks/PWGHF/taskLcK0sp.cxx

fix compilation

fixes

clang

chiarazampolli commented:
Last commit does not compile, but since the review has not started yet, it is no problem. I will try to fix it tomorrow, if I don't manage, I will revert.

chiarazampolli commented:
Now compiling.

chiarazampolli commented:
Hello @vkucera @ginnocen , 
Now also the histogramming is working. Missing is the MC part, but as we said, the review can start when you are ready (I will continue pushing here if I don't get comments). 
I have realized though that it would be better not to wait too long to merge this PR, since it involves also 2 HF classes that might change independently from my PR, which might then create troubles.

chiarazampolli commented:
Hello, 
As said on Tuesday, I change this PR to "ready for review", to verify that the CI does not complain. 
But then it should be moved back to "Draft" so that it is not merged accidentally by someone before @ginnocen and @vkucera review it. 
Chiara

chiarazampolli committed:
Avoid passing indices of daughters when building v0

In case, one would need a o2::dataformats::GlobalTrackID.h, but anyway it is not
used.

Fix

vkucera commented on Analysis/Tasks/PWGHF/HFTrackIndexSkimsCreator.cxx:
```diff
@@ -185,7 +438,7 @@ struct HFTrackIndexSkimsCreator {
      {"hmassDsToPiKK", "Ds candidates;inv. mass (K K #pi) (GeV/#it{c}^{2});entries", {HistType::kTH1F, {{500, 0., 5.}}}}}};
 
-  Filter filterSelectTracks = (aod::hf_seltrack::isSelProng > 0);
+  Filter filterSelectTracks = (aod::hf_seltrack::isSelProng > 0 || aod::hf_seltrack::isSelProng < 4);
```
Shouldn't this be `&&` instead of `||`?

chiarazampolli replied:
Yes

vkucera commented on Analysis/Tasks/PWGHF/HFTrackIndexSkimsCreator.cxx:
```diff
@@ -142,6 +173,228 @@ struct SelectTracks {
+        auto trackV0 = o2::dataformats::V0(vertexV0, momentumV0, posTrackParCov, negTrackParCov, {0, 0}, {0, 0}); // build the V0 track
+
+        // now we find the DCA between the V0 and the bachelor, for the cascade
+        int nCand2 = fitter.process(trackV0, bachTrack);
+
+        std::array<float, 3> pVecCandCasc = {0., 0., 0.};
+        if (nCand2 != 0) {
```
There is nothing happening after this block, so it can be
```c++
if (fitter.process(trackV0, bachTrack) == 0) {
  continue;
}
```

chiarazampolli replied:
done

vkucera commented on Analysis/Tasks/PWGHF/HFTrackIndexSkimsCreator.cxx:
```diff
@@ -142,6 +173,228 @@ struct SelectTracks {
+
+        std::array<float, 3> pVecCandCasc = {0., 0., 0.};
+        if (nCand2 != 0) {
+          fitter.propagateTracksToVertex();        // propagate the bach and V0 to the Lc vertex
```
Why is `propagateTracksToVertex` needed?

chiarazampolli replied:
To update the momentum

vkucera commented on Analysis/Tasks/PWGHF/HFTrackIndexSkimsCreator.cxx:
```diff
@@ -142,6 +173,228 @@ struct SelectTracks {
+          fitter.getTrack(0).getPxPyPzGlo(pvecv0); // take the momentum at the Lc vertex
+          fitter.getTrack(1).getPxPyPzGlo(pvecbach);
+
+          // HF code for the 2 prongs would use:
+          //auto pVecCandCasc = array{bachTrack.px() + trackV0.px(),
+          //			       bachTrack.py() + trackV0.py(),
+          //			       bachTrack.pz() + trackV0.pz()};
+          // but I think this is equivalent, and I understand it better where it comes from
+          pVecCandCasc = array{pvecbach[0] + pvecv0[0],
+                               pvecbach[1] + pvecv0[1],
+                               pvecbach[2] + pvecv0[2]};
+
+          // cascade candidate pT cut
+          if (RecoDecay::Pt(pVecCandCasc) < cutCascPtCandMin) {
```
For clarification: `bachTrack.px()` is momentum before vertexing, `pvecbach[0]` is after.
The momentum vector can be got as
```c++
auto pVecCandCasc = RecoDecay::PVec(pvecbach, pvecv0);
```
But you don't seem to need it now. So you could as well just call
```c++
RecoDecay::Pt(pvecbach, pvecv0) < cutCascPtCandMin
```
But maybe `pVecCandCasc` will be useful when there are more cuts.

chiarazampolli replied:
Yes, exactly. I removed the commented out code, it was from the very beginning of the development, when I was still finding my way in the procedure and vertexing.

vkucera commented on Analysis/Tasks/PWGHF/HFTrackIndexSkimsCreator.cxx:
```diff
@@ -142,6 +173,228 @@ struct SelectTracks {
+        const std::array<float, 3> vertexV0 = {v0.x(), v0.y(), v0.z()};
+        const std::array<float, 3> momentumV0 = {v0.px(), v0.py(), v0.pz()};
+
+        // we build the neutral track to then build the cascade
+        auto trackV0 = o2::dataformats::V0(vertexV0, momentumV0, posTrackParCov, negTrackParCov, {0, 0}, {0, 0}); // build the V0 track
```
I'm surprised that `{0, 0}` can be used instead of an integer index here.

chiarazampolli replied:
It is the ctor of the GlobalTrackID that is expected here

vkucera commented on Analysis/DataModel/include/AnalysisDataModel/HFSecondaryVertex.h:
```diff
@@ -299,6 +307,85 @@ DECLARE_SOA_TABLE(HfCandProng2MCRec, "AOD", "HFCANDP2MCREC",
+}
+
+template <typename T>
+auto InvMassGamma(const T& candidate)
```
I'm not sure that `InvMassGamma` should be in `hf_cand_casc` as conversion is not a cascade.
`InvMassGammaToEE`

chiarazampolli replied:
I thought (just from reading the code) that here I should put all the things that are needed for the cascade analysis. Correct me if I am wrong.

vkucera commented on Analysis/DataModel/include/AnalysisDataModel/HFSecondaryVertex.h:
```diff
@@ -299,6 +307,85 @@ DECLARE_SOA_TABLE(HfCandProng2MCRec, "AOD", "HFCANDP2MCREC",
+
+namespace hf_cand_casc
+{
+DECLARE_SOA_EXPRESSION_COLUMN(Px, px, float, 1.f * aod::hf_cand::pxProng0 + 1.f * aod::hf_cand::pxProng1);
+DECLARE_SOA_EXPRESSION_COLUMN(Py, py, float, 1.f * aod::hf_cand::pyProng0 + 1.f * aod::hf_cand::pyProng1);
+DECLARE_SOA_EXPRESSION_COLUMN(Pz, pz, float, 1.f * aod::hf_cand::pzProng0 + 1.f * aod::hf_cand::pzProng1);
+DECLARE_SOA_DYNAMIC_COLUMN(M, m, [](float px0, float py0, float pz0, float px1, float py1, float pz1, const array<double, 2>& m) { return RecoDecay::M(array{array{px0, py0, pz0}, array{px1, py1, pz1}}, m); });
```
Why are these needed? Cannot you just create `HfCandCascExt` with the `hf_cand_prong2` columns?

chiarazampolli replied:
No it does not work, I had a discussion with @aalkin for px, py, pz (not the mass though) since it was not compiling, and we concluded that I had to create my own columns.

vkucera commented on Analysis/Tasks/PWGHF/CMakeLists.txt:
```diff
@@ -63,6 +68,11 @@ o2_add_dpl_workflow(hf-jpsi-toee-candidate-selector
                     PUBLIC_LINK_LIBRARIES O2::Framework O2::AnalysisDataModel O2::AnalysisCore O2::DetectorsVertexing
                     COMPONENT_NAME Analysis)
 
+o2_add_dpl_workflow(hf-lck0sp-candidate-selector
+                    SOURCES HFLcK0spCandidateSelector.cxx
```
`hf-lc-tok0sp-candidate-selector`
`HFLcToK0sPCandidateSelector`

chiarazampolli replied:
done

vkucera commented on Analysis/Tasks/PWGHF/HFCandidateCreatorCascade.cxx:
```diff
@@ -0,0 +1,211 @@
+
+  Produces<aod::HfCandCascBase> rowCandidateBase;
+
+  Configurable<double> d_bz{"d_bz", 5., "magnetic field"};
+  Configurable<bool> b_propdca{"b_propdca", true, "create tracks version propagated to PCA"};
+  Configurable<double> d_maxr{"d_maxr", 200., "reject PCA's above this radius"};
+  Configurable<double> d_maxdzini{"d_maxdzini", 4., "reject (if>0) PCA candidate if tracks DZ exceeds threshold"};
+  Configurable<double> d_minparamchange{"d_minparamchange", 1.e-3, "stop iterations if largest change of any X is smaller than this"};
+  Configurable<double> d_minrelchi2change{"d_minrelchi2change", 0.9, "stop iterations is chi2/chi2old > this"};
+  Configurable<bool> b_dovalplots{"b_dovalplots", true, "do validation plots"};
```
Please change these (and anything else) to the camelCase convention. It will make the renaming campaign easier. :-)

chiarazampolli replied:
I hope I caught them all. In the IndexSkimCreator, there are other parts where this should be done, should I do it?

vkucera commented on Analysis/DataModel/include/AnalysisDataModel/HFCandidateSelectionTables.h:
```diff
@@ -19,24 +19,25 @@ DECLARE_SOA_COLUMN(IsSelD0, isSelD0, int);
+
+namespace hf_selcandidate_lc_k0sp
+{
+DECLARE_SOA_COLUMN(IsSelLcK0sp, isSelLcK0sp, int);
+} // namespace hf_selcandidate_lc_k0sp
+DECLARE_SOA_TABLE(HFSelLcK0spCandidate, "AOD", "HFSELLCK0SPCAND", hf_selcandidate_lc_k0sp::IsSelLcK0sp);
```
`IsSelLcToK0sP, isSelLcToK0sP`
`HFSelLcToK0sPCandidate`

chiarazampolli replied:
done

vkucera commented on Analysis/DataModel/include/AnalysisDataModel/HFSecondaryVertex.h:
```diff
@@ -38,6 +39,7 @@ DECLARE_SOA_TABLE(HFSelTrack, "AOD", "HFSELTRACK",
 using BigTracks = soa::Join<Tracks, TracksCov, TracksExtra, HFSelTrack>;
 using BigTracksMC = soa::Join<BigTracks, McTrackLabels>;
 using BigTracksPID = soa::Join<BigTracks, pidRespTPC, pidRespTOF>;
+using V0DataExt = aod::V0DataExt;
```
Is this alias used anywhere?

chiarazampolli replied:
Yes, line 52

vkucera replied:
Well, an alias that is used just once is not a very useful one, is it? :-)

vkucera replied:
It is called `V0Datas` now but my comment still applies. This alias is used only once and you use `aod::V0Datas` everywhere else. So I don't see any benefit of defining this alias at all.

chiarazampolli replied:
I don't think (or know) that it makes any overkill, but ok, since you don't like it, I removed it. I introduced it when I did not know how many times I would have needed it.

vkucera commented on Analysis/DataModel/include/AnalysisDataModel/HFSecondaryVertex.h:
```diff
@@ -299,6 +307,85 @@ DECLARE_SOA_TABLE(HfCandProng2MCRec, "AOD", "HFCANDP2MCREC",
+DECLARE_SOA_EXPRESSION_COLUMN(Pz, pz, float, 1.f * aod::hf_cand::pzProng0 + 1.f * aod::hf_cand::pzProng1);
+DECLARE_SOA_DYNAMIC_COLUMN(M, m, [](float px0, float py0, float pz0, float px1, float py1, float pz1, const array<double, 2>& m) { return RecoDecay::M(array{array{px0, py0, pz0}, array{px1, py1, pz1}}, m); });
+
+template <typename T>
+auto InvMassLc(const T& candidate)
```
`InvMassLcToK0sP`

chiarazampolli replied:
done

vkucera commented on Analysis/Tasks/PWGHF/HFCandidateCreatorCascade.cxx:
```diff
@@ -0,0 +1,211 @@
+  OutputObj<TH1F> hCovPVXX{TH1F("hCovPVXX", "2-prong candidates;XX element of cov. matrix of prim. vtx. position (cm^{2});entries", 100, 0., 1.e-4)};
+  OutputObj<TH1F> hCovSVXX{TH1F("hCovSVXX", "2-prong candidates;XX element of cov. matrix of sec. vtx. position (cm^{2});entries", 100, 0., 0.2)};
+
+  double massP = RecoDecay::getMassPDG(kProton);
+  double massK0s = RecoDecay::getMassPDG(kK0Short);
+  double massPi = RecoDecay::getMassPDG(kPiPlus);
+  double massLc = RecoDecay::getMassPDG(4122);
```
Maybe these could be `const`.

chiarazampolli replied:
done

chiarazampolli replied:
Actually, it cannot be const:
```
/home/zampolli/SOFT/alibuild/ali-o2-analysis/sw/SOURCES/O2/dev/0/Analysis/Tasks/PWGHF/HFCandidateCreatorCascade.cxx: In member function ‘void HFCandidateCreatorCascade::process(const Collisions&, const HfTrackIndexCasc&, const BigTracks&, const V0DataExt&)’:
/home/zampolli/SOFT/alibuild/ali-o2-analysis/sw/SOURCES/O2/dev/0/Analysis/Tasks/PWGHF/HFCandidateCreatorCascade.cxx:145:19: error: assignment of read-only member ‘HFCandidateCreatorCascade::mass2K0sP’
  145 |         mass2K0sP = RecoDecay::M(arrayMomenta, array{massK0s, massP});
```

vkucera replied:
I didn't mention `mass2K0sP` in my suggestion.

vkucera replied:
My comment still applies.

vkucera commented on Analysis/DataModel/include/AnalysisDataModel/HFSecondaryVertex.h:
```diff
@@ -299,6 +307,85 @@ DECLARE_SOA_TABLE(HfCandProng2MCRec, "AOD", "HFCANDP2MCREC",
+                  hf_cand::ImpactParameter0, hf_cand::ImpactParameter1,
+                  hf_cand::ErrorImpactParameter0, hf_cand::ErrorImpactParameter1,
+                  hf_track_index::IndexV0Id, // V0 index
+                  hf_track_index::Index0Id,
```
Shouldn't the bachelor have `Index1Id` if everywhere the V0 goes first and the bachelor second?

chiarazampolli replied:
The V0 has index IndexV0Id, the bachelor is the first track of the candidate, so Index0Id. I would leave it as it is.

vkucera commented on Analysis/Tasks/PWGHF/HFLcK0spCandidateSelector.cxx:
```diff
@@ -0,0 +1,336 @@
+  template <typename T>
+  bool validTPCPID(const T& track)
+  {
+    if (TMath::Abs(track.pt()) < d_pidTPCMinpT || TMath::Abs(track.pt()) >= d_pidTPCMaxpT) {
```
This doesn't make sense to me. Why would `pt` be negative?
@nzardosh This is already in `HFD0CandidateSelector`. Why?

chiarazampolli replied:
I had the same doubt, but thought that this was already thought over

vkucera replied:
It's a bug. Please fix it.

chiarazampolli replied:
done. It is not a bug in the true sense, but clearly it is useless and introduces an overhead.

vkucera commented on Analysis/Tasks/PWGHF/HFLcK0spCandidateSelector.cxx:
```diff
@@ -0,0 +1,336 @@
+  {
+    double nSigma = 100.0; //arbitarily large value
+    nSigma = track.tpcNSigmaPr();
+    return nSigma > nSigmaCut;
```
`nSigma < nSigmaCut`
@nzardosh Shouldn't here be `std::abs(nSigma)`?

chiarazampolli replied:
I will add it, because I agree

vkucera commented on Analysis/Tasks/PWGHF/HFTrackIndexSkimsCreator.cxx:
```diff
@@ -21,8 +21,13 @@
 #include "AnalysisCore/trackUtilities.h"
 #include "AnalysisCore/HFConfigurables.h"
-//#include "AnalysisDataModel/Centrality.h"
 #include "Framework/HistogramRegistry.h"
```
Why?

chiarazampolli replied:
Historical reason... Initially I needed to run some v0 workflow to get some info and this v0 workflow needed centrality. Will remove it.

chiarazampolli replied:
But this is not in the code anymore...

vkucera replied:
The point is that the related pieces are still in the code, just commented out for later. Your PR removes this line, but it should stay there.

vkucera replied:
My comment still applies.

chiarazampolli replied:
Re-added line.

vkucera commented on Analysis/Tasks/PWGHF/HFTrackIndexSkimsCreator.cxx:
```diff
@@ -52,7 +57,12 @@ struct SelectTracks {
+  Configurable<double> dcatoprimxy_maxpt_bach{"dcatoprimxy_maxpt_bach", 2., "max pt cut for min. DCAXY to prim. vtx. for bachelor in cascade candidate"};
+  Configurable<double> dcatoprimxymin_bach{"dcatoprimxymin_bach", 0., "min. DCAXY to prim. vtx. for bachelor in cascade candidate"}; // for PbPb 2018, the cut should be 0.0025
+  Configurable<double> dcatoprimxymax_bach{"dcatoprimxymax_bach", 1.0, "max. DCAXY to prim. vtx. for bachelor in cascade candidate"};
+  Configurable<double> etamax_bach{"etamax_bach", 0.8, "max. pseudorapidity for bachelor in cascade candidate"};
   HistogramRegistry registry{
```
Please keep a blank line here.

chiarazampolli replied:
done

vkucera commented on Analysis/Tasks/PWGHF/HFTrackIndexSkimsCreator.cxx:
```diff
@@ -94,11 +113,14 @@ struct SelectTracks {
       if (doCutQuality.value && status_prong > 0) { // FIXME to make a more complete selection e.g track.flags() & o2::aod::track::TPCrefit && track.flags() & o2::aod::track::GoldenChi2 &&
         UChar_t clustermap = track.itsClusterMap();
-        if (!(track.tpcNClsFound() >= d_tpcnclsfound.value &&
+        if (!(track.tpcNClsFound() >= d_tpcnclsfound.value && // is this the number of TPC clusters? It should not be used
```
If it should not be used, it should be removed. It should be checked.

chiarazampolli replied:
This is in several tasks, I wanted that we all are on the same page. We can discuss this at the next meeting (next week, since tomorrow's is cancelled).

vkucera commented on Analysis/Tasks/PWGHF/HFTrackIndexSkimsCreator.cxx:
```diff
@@ -142,6 +173,228 @@ struct SelectTracks {
+/// to run: o2-analysis-weak-decay-indices --aod-file AO2D.root -b | o2-analysis-lambdakzerobuilder -b |
+///         o2-analysis-trackextension -b | o2-analysis-hf-track-index-skims-creator -b
+
+struct HFTrackIndexSkimsCreatorCascades {
```
I would prefer to put it after `HFTrackIndexSkimsCreator`, not before.

chiarazampolli replied:
sure, it is before for no reason - I copied-pasted, and started from the copy at the top to be closer to the selection for practical reason.

vkucera commented on Analysis/Tasks/PWGHF/HFTrackIndexSkimsCreator.cxx:
```diff
@@ -142,6 +173,228 @@ struct SelectTracks {
+  using MyTracks = soa::Join<aod::FullTracks, aod::HFSelTrack, aod::TracksExtended>;
+  //Partition<MyTracks> selectedTracks = aod::hf_seltrack::isSelProng >= 4;
+  // using SelectedV0s = soa::Filtered<aod::V0DataExt>;
+  double massP = RecoDecay::getMassPDG(kProton);
+  double massK0s = RecoDecay::getMassPDG(kK0Short);
+  double massPi = RecoDecay::getMassPDG(kPiPlus);
+  double massLc = RecoDecay::getMassPDG(4122);
```
`const`?

chiarazampolli replied:
done

chiarazampolli replied:
Can't be const actually:
```
/home/zampolli/SOFT/alibuild/ali-o2-analysis/sw/SOURCES/O2/dev/0/Analysis/Tasks/PWGHF/HFCandidateCreatorCascade.cxx: In member function ‘void HFCandidateCreatorCascade::process(const Collisions&, const HfTrackIndexCasc&, const BigTracks&, const V0DataExt&)’:
/home/zampolli/SOFT/alibuild/ali-o2-analysis/sw/SOURCES/O2/dev/0/Analysis/Tasks/PWGHF/HFCandidateCreatorCascade.cxx:145:19: error: assignment of read-only member ‘HFCandidateCreatorCascade::mass2K0sP’
  145 |         mass2K0sP = RecoDecay::M(arrayMomenta, array{massK0s, massP});
```
(from the other class, but it is the same here)

vkucera replied:
My comment still applies.

chiarazampolli replied:
I don't understand the comment. I cannot make them `const`.

vkucera commented on Analysis/Tasks/PWGHF/HFTrackIndexSkimsCreator.cxx:
```diff
@@ -142,6 +173,228 @@ struct SelectTracks {
+          // calculate invariant masses
+          auto arrMom = array{pvecbach, pvecv0};
+          mass2K0sP = RecoDecay::M(arrMom, array{massP, massK0s});
+          // invariant-mass cut
+          if (cutCascInvMassLc >= 0.) {
+            if (std::abs(mass2K0sP - massLc) < cutCascInvMassLc) {
+              continue;
+            }
+          }
```
The two `if`s can be merged.
For 2-prong skims, the mass cut is applied before vertexing. Is there a reason to do it after in this case?

chiarazampolli replied:
I don't understand: how can I do it before? I need the momenta at the secondary vertex. Maybe I am overlooking something...

chiarazampolli replied:
BTW, now there is 1 if. Not sure it has any impact though...

vkucera commented on Analysis/Tasks/PWGHF/taskLcK0sp.cxx:
```diff
@@ -0,0 +1,157 @@
+#include "Framework/runDataProcessing.h"
+
+/// LcK0sp analysis task
+struct TaskLcK0sp {
```
`TaskLcToK0sP`

chiarazampolli replied:
But the symbol of the proton is small "p". I understand that you (or O2?) have the camelCase rule, but I would leave the proton small. No?

chiarazampolli replied:
done, better to follow the rules.

vkucera commented on Analysis/Tasks/PWGHF/taskLcK0sp.cxx:
```diff
@@ -0,0 +1,157 @@
+        continue;
+      }
+      */
+      if (cutEtaCandMax >= 0. && std::abs(candidate.eta()) > cutEtaCandMax) {
+        //Printf("Candidate: eta rejection: %g", candidate.eta());
```
I suppose this will become a rapidity cut eventually.

chiarazampolli replied:
Yes. For now, I uniformized with D0 and Lc --> pKpi

vkucera commented on Analysis/Tasks/PWGHF/HFTrackIndexSkimsCreator.cxx:
```diff
@@ -21,8 +21,13 @@
 #include "AnalysisCore/HFConfigurables.h"
-//#include "AnalysisDataModel/Centrality.h"
 #include "Framework/HistogramRegistry.h"
+#include "AnalysisDataModel/StrangenessTables.h"
+#include "AnalysisDataModel/TrackSelectionTables.h"
+#include "ReconstructionDataFormats/V0.h"
+#include "Framework/ASoAHelpers.h"
+#include "Framework/AnalysisDataModel.h"
```
Are all these needed?

chiarazampolli replied:
Removed useless ones.

vkucera commented on Analysis/Tasks/PWGHF/CMakeLists.txt:
```diff
@@ -87,3 +97,8 @@ o2_add_dpl_workflow(hf-task-bplus
                     PUBLIC_LINK_LIBRARIES O2::Framework O2::AnalysisDataModel O2::AnalysisCore O2::DetectorsVertexing
                     COMPONENT_NAME Analysis)
+
+o2_add_dpl_workflow(hf-task-lck0sp
+                    SOURCES taskLcK0sp.cxx
```
`taskLcToK0sP`

chiarazampolli replied:
See above...

chiarazampolli replied:
done

vkucera reviewed (commented):
Hi @chiarazampolli , a couple of first comments.
In general:
- Fix naming (camelCase).
- Add missing readability braces around statements.

chiarazampolli committed:
Comments by Vit

clang

Fixes for braces

clang

chiarazampolli committed:
Adding pT V0 daugh, and swapping V0 and bach (as in AliPhysics)

clang

chiarazampolli commented:
The error in the CI (which is now re-running because I added a further commit) seems to be coming from CPV:
```
Launching task: /mnt/mesos/sandbox/sandbox/o2-fullci/sw/slc8_x86-64/O2/5257-1/prodtests/full-system-test/dpl-workflow.sh &> reco_NOGPU.log &
Detected critical problem in logfile reco_NOGPU.log
reco_NOGPU.log-[61732:mft-tracker]: [14:29:12][INFO] Found tracks CA: 0
reco_NOGPU.log-[61732:mft-tracker]: [14:29:12][INFO] ROframe: 9, clusters loaded : 28765
reco_NOGPU.log:[61858:CPVClusterizerSpec]: [14:29:12][ERROR] Exception caught: bitset::test: __position (which is 23040) >= _Nb (which is 23040) 
reco_NOGPU.log-[61858:CPVClusterizerSpec]: /mnt/mesos/sandbox/sandbox/o2-fullci/sw/slc8_x86-64/O2/5257-1/lib/libO2FrameworkFoundation.so(_ZN2o29framework13runtime_errorEPKc+0x74)[0x7f26d3145a54]
reco_NOGPU.log-[61858:CPVClusterizerSpec]: /mnt/mesos/sandbox/sandbox/o2-fullci/sw/slc8_x86-64/O2/5257-1/lib/libO2Framework.so(+0x1106a5)[0x7f26d425f6a5]
```
Pinging @peressounko , @shahor02 . 

Chiara

peressounko commented:
Dear @chiarazampolli CPV bug should be fixed in PR https://github.com/AliceO2Group/AliceO2/pull/5499

chiarazampolli commented:
Hi @peressounko ,

Thanks, indeed the CI is now green!

Chiara

chiarazampolli committed:
First version using MC info.

To be investigated:
- why no V0 in the V0 table is a K0s that comes from a Lc?
- the RecoDecay::isMatchedMCGen might need some adjustments, being checked

Fixes

Format

ktf commented:
Looks like there is a conflict now...

chiarazampolli commented:
Yes, I am still working on some MC part, I will fix everything when I am done.

vkucera commented:
@ginnocen Can you convert this back to draft please to avoid triggering the full tests with every new push?
This is still development in progress.

ginnocen commented:
Done :) @chiarazampolli

chiarazampolli committed:
recent work - to be cleaned-up

chiarazampolli committed:
Update Lc -> K0S+p to do MC checks too.

Adding also the possibility to do local debugging with an AO2D file from
any production.

format

Stupid fix

chiarazampolli commented:
Hello @vkucera ,

For me now the PR is complete. I added all ingredients, and verified step by step that things work. You will see that I implemented a way to do easy local debugging. 
Please, start reviewing the PR. I propose that we do not postpone in case there are major updates, since the PR is now quite big. Further important changes should go to different PR's. 

Thanks!

Chiara

chiarazampolli commented:
As last time, note that the PR is not ready to be merged, but as I wrote it is ready for review. I know that  in PWGHF it is preferred to keep PRs Draft before the final step to be merged, having it not Draft allows the tests to be run. Since I introduced several changes, this is something useful.

chiarazampolli commented:
Dear @jgrosseo , @ktf , @aalkin , @ginnocen 

After a discussion with @vkucera , we would like to get your opinion about this PR, namely on the fact that I have implemented a way (through an `IFDEF`) to do some local debugging of my code, which introduces also an external library (a tini-tiny one). The doubt of @vkucera is that the library is case-specific. 
I don't think that it hurts in any way, and one could think of using an ENV variable, or something, to pass the values that are code-specific, but it is surely better if you could take a look. See https://github.com/AliceO2Group/AliceO2/pull/5257/files#diff-dbd3d9a608665d615cae0f8dce5ca74b7d31350221def9bbe7e49c598d9bfd91.

Thanks!!

Chiara

chiarazampolli commented:
Dear @jgrosseo , @aalkin , @ktf , @ginnocen ,

I know that you are super busy, but this is a ping if you could check this PR... Not the details, just the specific implementation discussed in the comment above. 

Thanks!

Chiara

ktf commented:
There are some conflicts. Could you fix them?

ktf commented on Analysis/Tasks/PWGLF/lambdakzerobuilder.cxx:
```diff
@@ -73,6 +74,19 @@ DECLARE_SOA_TABLE(V0GoodIndices, "AOD", "V0GOODINDICES", o2::soa::Index<>,
+//#define MY_DEBUG
+#ifdef MY_DEBUG
+using MyTracks = FullTracksExtMC;
+#define MY_DEBUG_MSG(condition, cmd) \
```
There is `LOG_IF(severity, condition)` which you can use. Something like:

LOF_IF(error, a < b) << "a < b!!!!!";

chiarazampolli replied:
Are you suggesting to replace 
```
#define MY_DEBUG_MSG(condition, cmd)
```
with 
```
#define LOG_IF(severity, condition)
```
Because in my implementation (suggested by David, to be honest), the MY_DEBUG_MSG changes behaviour according to whether there is or not the MY_DEBUG defined.

ktf commented on Analysis/Tasks/PWGLF/lambdakzerobuilder.cxx:
```diff
@@ -93,36 +107,56 @@ struct lambdakzeroprefilterpairs {
+               MyTracks const& tracks
+#ifdef MY_DEBUG
+               ,
+               aod::McParticles const& particlesMC
```
@aalkin We need to add `Lifetime::Optional` support to the analysis framework.

chiarazampolli commented:
> There are some conflicts. Could you fix them?


Ciao @ktf , 

Thanks for the feedback.

I don't see the conflict now, but for sure there will be after a PR by Jan Fiete. The PR is still in Draft mode, to not have it tested before the review (this is how it is done in HF). Maybe the CI test will show the conflict (will it run now? It is in "restarting" mode, I will check again tomorrow/Monday), then I will fix it. But the point was more to give feedback to the implementation with the extra library and the IFDEF. 

I take from your comments that you don't see a problem in the implementation, right? - apart from the suggestion to use the LOG_IF, which I am not sure I can use. I will add a comment where you replied about it. 

Chiara

vkucera commented:
@jgrosseo , @ktf , @aalkin , @ginnocen
The main issues I have with `MY_DEBUG` are:
* `UtilsDebugLcK0Sp.h` is a list of hard-coded conditions that work only with one specific input file.
* It is not documented anywhere how to get these conditions for another input.
* It is not clear to me what it is supposed to debug. The functions in `UtilsDebugLcK0Sp.h` are related to verification of correct MC matching, which relies on functions in `RecoDecay` already used in several other HF tasks. As I understood, Chiara has tested that this matching works in her code and therefore I don't see a reason for the debugging code to be committed.

vkucera commented on Analysis/Core/include/AnalysisCore/RecoDecay.h:
```diff
@@ -707,7 +707,7 @@ class RecoDecay
         // Check that the number of direct daughters is not larger than the number of expected final daughters.
         if (indexDaughterFirst > -1 && indexDaughterLast > -1 && indexDaughterLast - indexDaughterFirst + 1 > N) {
-          //Printf("MC Rec: Rejected: too many direct daughters: %d (expected %d final)", indexDaughterLast - indexDaughterFirst + 1, N);
+          //Printf("MC Rec: Rejected: too many direct daughters: %d (expected %ld final)", indexDaughterLast - indexDaughterFirst + 1, N);
```
What is the motivation of these changes?

chiarazampolli replied:
Compilation warnings when uncommenting the lines.

chiarazampolli commented:
> @jgrosseo , @ktf , @aalkin , @ginnocen
> The main issues I have with `MY_DEBUG` are:
> 
>     * `UtilsDebugLcK0Sp.h` is a list of hard-coded conditions that work only with one specific input file.
> 
>     * It is not documented anywhere how to get these conditions for another input.
> 
>     * It is not clear to me what it is supposed to debug. The functions in `UtilsDebugLcK0Sp.h` are related to verification of correct MC matching, which relies on functions in `RecoDecay` already used in several other HF tasks. As I understood, Chiara has tested that this matching works in her code and therefore I don't see a reason for the debugging code to be committed.

Ciao @vkucera , @jgrosseo , @ktf , @aalkin , @ginnocen

What Vit writes is true, the `UtilsDebugLcK0Sp.h` contains checks that are specific to a file I used for debugging. Still, I think it is not an issue to have it, considering that:
- I can add documentation, and the macro I used - which is indeed a good idea, the same approach could be useful for others.
- it is super tiny, it will not harm anyone; and throughout O2 we do have macros and things for checks and debug
- it is meant to check what happens throughout the code to the few Lc candidates without going through the decay-tree checks "if they are really a Lc". I can check how to do the same with the functions in `RecoDecay` (which I myself use somewhere else too), but this is just a quick way to do what I need for debugging. I would like to keep this piece of work, which made it easier for me to implement the code, and which can make it easier also in the future. 

Thanks for any suggestions, and to Vit for his huge patience.

Chiara

PS for Vit: I will go through the other comments in the next days.

vkucera commented on Analysis/Tasks/PWGHF/HFTrackIndexSkimsCreator.cxx:
```diff
@@ -851,9 +931,294 @@ struct HFTrackIndexSkimsCreator {
+  }   // process
+};
+
+struct HFTrackIndexTestMCCasc {
+  void process(aod::BigTracksMC const& tracks,
+               aod::V0Datas const& V0s,
+               aod::McParticles const& particlesMC)
```
What is this device supposed to do? It seems to just loop and match without any output.

chiarazampolli replied:
It was used to test the MC matching for the K0S daughters. A sort of debug task. I will remove it.

vkucera commented on Analysis/Tasks/PWGHF/HFTrackIndexSkimsCreator.cxx:
```diff
@@ -851,9 +931,294 @@ struct HFTrackIndexSkimsCreator {
     adaptAnalysisTask<SelectTracks>(cfgc, "hf-produce-sel-track"),
-    adaptAnalysisTask<HFTrackIndexSkimsCreator>(cfgc, "hf-track-index-skims-creator")};
+    adaptAnalysisTask<HFTrackIndexSkimsCreator>(cfgc, "hf-track-index-skims-creator"),
+    adaptAnalysisTask<HFTrackIndexSkimsCreatorCascades>(cfgc, "hf-track-index-skims-cascades-creator")};
```
I think the cascade skimming should be optional. Otherwise it always runs, even when only the standard 2/3-prong skimming is wanted and one has to run `lambdakzerobuilder` to produce `V0Datas`.

chiarazampolli replied:
Done

jgrosseo commented:
> 
> 
> > @jgrosseo , @ktf , @aalkin , @ginnocen
> > The main issues I have with `MY_DEBUG` are:
> > ```
> > * `UtilsDebugLcK0Sp.h` is a list of hard-coded conditions that work only with one specific input file.
> > 
> > * It is not documented anywhere how to get these conditions for another input.
> > 
> > * It is not clear to me what it is supposed to debug. The functions in `UtilsDebugLcK0Sp.h` are related to verification of correct MC matching, which relies on functions in `RecoDecay` already used in several other HF tasks. As I understood, Chiara has tested that this matching works in her code and therefore I don't see a reason for the debugging code to be committed.
> > ```
> 
> Ciao @vkucera , @jgrosseo , @ktf , @aalkin , @ginnocen
> 
> What Vit writes is true, the `UtilsDebugLcK0Sp.h` contains checks that are specific to a file I used for debugging. Still, I think it is not an issue to have it, considering that:
> 
>     * I can add documentation, and the macro I used - which is indeed a good idea, the same approach could be useful for others.
> 
>     * it is super tiny, it will not harm anyone; and throughout O2 we do have macros and things for checks and debug
> 
>     * it is meant to check what happens throughout the code to the few Lc candidates without going through the decay-tree checks "if they are really a Lc". I can check how to do the same with the functions in `RecoDecay` (which I myself use somewhere else too), but this is just a quick way to do what I need for debugging. I would like to keep this piece of work, which made it easier for me to implement the code, and which can make it easier also in the future.
> 
> 
> Thanks for any suggestions, and to Vit for his huge patience.
> 
> Chiara
> 
> PS for Vit: I will go through the other comments in the next days.

I would say that specific checks (which run only with a file which is not in the repository) should then be in macros, but not in compiled header files, so that we do not add that the compiled code size.

The correct way would be to implement a test which also populates the tables, like there exist for the Framework tests.

chiarazampolli committed:
Revisiting debugging

chiarazampolli commented:
Hello @aalkin , @vkucera ,
The last commit implements what we discussed during the very useful chat today, with the exception that the function is still in an external library, since I need it in several tasks. 
What do you think?
Chiara

aalkin commented on Analysis/Tasks/PWGHF/HFCandidateCreatorCascade.cxx:
```diff
@@ -44,7 +44,12 @@ struct HFCandidateCreatorCascade {
+#ifdef MY_DEBUG
+  Configurable<std::vector<int>> v_labelK0Spos{"v_labelK0Spos", {729, 2866, 4754}, "labels of K0S positive daughters, for debug"};
+  Configurable<std::vector<int>> v_labelK0Sneg{"v_labelK0Sneg", {730, 2867, 4755}, "labels of K0S positive daughters, for debug"};
+  Configurable<std::vector<int>> v_labelProton{"v_labelProton", {729, 2810, 4393}, "labels of K0S positive daughters, for debug"};
```
Minor thing: the help strings should be different.

chiarazampolli replied:
Of course.

aalkin commented on Analysis/Tasks/Utils/include/AnalysisTasksUtils/UtilsDebugLcK0Sp.h:
```diff
@@ -10,27 +10,54 @@
 {
 
-  return (labelK0SPos == 729 && labelK0SNeg == 730) || (labelK0SPos == 2866 && labelK0SNeg == 2867) || (labelK0SPos == 4754 && labelK0SNeg == 4755) || (labelK0SPos == 5457 && labelK0SNeg == 5458) || (labelK0SPos == 6891 && labelK0SNeg == 6892) || (labelK0SPos == 7824 && labelK0SNeg == 7825) || (labelK0SPos == 9243 && labelK0SNeg == 9244) || (labelK0SPos == 9810 && labelK0SNeg == 9811) || (labelK0 ...
+  assert(listLabelsK0SPos.size() == listLabelsK0SPos.size());
```
Assertions only have effect when compiled in debug mode, either have a normal `if` and throw an exception if it fails, or remove it.

chiarazampolli replied:
Done

aalkin commented on Analysis/Tasks/Utils/include/AnalysisTasksUtils/UtilsDebugLcK0Sp.h:
```diff
@@ -10,27 +10,54 @@
 
-  return (labelK0SPos == 729 && labelK0SNeg == 730) || (labelK0SPos == 2866 && labelK0SNeg == 2867) || (labelK0SPos == 4754 && labelK0SNeg == 4755) || (labelK0SPos == 5457 && labelK0SNeg == 5458) || (labelK0SPos == 6891 && labelK0SNeg == 6892) || (labelK0SPos == 7824 && labelK0SNeg == 7825) || (labelK0SPos == 9243 && labelK0SNeg == 9244) || (labelK0SPos == 9810 && labelK0SNeg == 9811) || (labelK0 ...
+  assert(listLabelsK0SPos.size() == listLabelsK0SPos.size());
+  for (auto i = 0; i < listLabelsK0SPos.size(); ++i) {
```
Consider using `std::any_of` instead of a for loop.

chiarazampolli replied:
Done

chiarazampolli committed:
Anton's comments

clang-format

chiarazampolli commented:
Hello @aalkin , @vkucera , 

I implemented Anton's comments. Do you have more on the debugging? If not, I would then move on with the remaining old ones from Vit, plus rebasing.

Thanks!

Chiara

aalkin commented on Analysis/Tasks/PWGHF/HFCandidateCreatorCascade.cxx:
```diff
@@ -0,0 +1,301 @@
+  double massP = RecoDecay::getMassPDG(kProton);
+  double massK0s = RecoDecay::getMassPDG(kK0Short);
+  double massPi = RecoDecay::getMassPDG(kPiPlus);
+  double massLc = RecoDecay::getMassPDG(4122);
```
Avoid using PDG codes directly, there is now `HFSelectorCuts.h` header that includes enum with HF PDG codes.

chiarazampolli replied:
Done.

aalkin commented on Analysis/Tasks/PWGHF/HFLcK0sPCandidateSelector.cxx:
```diff
@@ -0,0 +1,400 @@
+static const int nCutVars = 8;
+//temporary until 2D array in configurable is solved - then move to json
+//mK0s(GeV)     mLambdas(GeV)    mGammas(GeV)    ptp     ptK0sdau     pTLc     d0p     d0K0
+constexpr double cuts[nPtBins][nCutVars] = {{0.008, 0.005, 0.1, 0.5, 0.3, 0.6, 0.05, 999999.},  // 1 < pt < 2
```
This would need to be turned into a configurable, like in other HF tasks.

chiarazampolli replied:
Yes, but I would leave this to another PR. This is already big, and I might not have time to work on it in the next couple of weeks.

aalkin replied:
Sure, that works.

chiarazampolli committed:
Implementation of comments by Anton and Vit

misc fixes

clang-format

chiarazampolli commented:
Hello @vkucera , @aalkin ,
I rebased, and added changes after your comments. 
Can you take a look?
Thanks!
Cheers,
Chiara

vkucera commented on Analysis/Tasks/PWGHF/HFTrackIndexSkimsCreator.cxx:
```diff
@@ -20,10 +20,12 @@
 #include "AnalysisDataModel/TrackSelectionTables.h"
 #include "ReconstructionDataFormats/V0.h"
 #include "AnalysisTasksUtils/UtilsDebugLcK0Sp.h"
+#include "AnalysisCore/HFSelectorCuts.h"
```
Not needed (included in `HFSecondaryVertex.h`).

chiarazampolli replied:
Removed

vkucera commented on Analysis/Tasks/PWGHF/HFTrackIndexSkimsCreator.cxx:
```diff
@@ -32,10 +34,11 @@ using namespace o2::framework;
 using namespace o2::framework::expressions;
 using namespace o2::aod::hf_cand_prong2;
 using namespace o2::aod::hf_cand_prong3;
+using namespace o2::analysis;
```
Not needed (declared in `HFSecondaryVertex.h`).

vkucera commented on Analysis/Tasks/PWGHF/taskLcK0sP.cxx:
```diff
@@ -19,11 +19,13 @@
 #include "Framework/HistogramRegistry.h"
 #include "AnalysisDataModel/HFSecondaryVertex.h"
 #include "AnalysisDataModel/HFCandidateSelectionTables.h"
+#include "AnalysisCore/HFSelectorCuts.h"
```
Not needed (included in `HFSecondaryVertex.h`).

vkucera commented on Analysis/Tasks/PWGHF/taskLcK0sP.cxx:
```diff
@@ -19,11 +19,13 @@
 using namespace o2::framework;
 using namespace o2::aod::hf_cand_casc;
 using namespace o2::framework::expressions;
+using namespace o2::analysis;
```
Not needed (declared in `HFSecondaryVertex.h`).

vkucera commented on Analysis/Core/include/AnalysisCore/RecoDecay.h:
```diff
@@ -874,7 +874,7 @@ class RecoDecay
       }
     }
-    //Printf("MC Gen: Accepted: m: %d", candidate.globalIndex());
+    //Printf("MC Gen: Accepted: m: %ld", candidate.globalIndex());
```
Does this one really trigger a warning? Index should be `int`.

chiarazampolli replied:
I don't recall, I changed it back to int. One will see when (s)he uncomments the line, it might be that I changed it by mistake.

vkucera commented on Analysis/DataModel/include/AnalysisDataModel/HFSecondaryVertex.h:
```diff
@@ -311,6 +318,100 @@ DECLARE_SOA_TABLE(HfCandProng2MCGen, "AOD", "HFCANDP2MCGEN",
+                  hf_track_index::IndexV0Id, // V0 index
+                  hf_track_index::HFflag,
+                  // V0
+                  v0data::X, v0data::Y, v0data::Z,
+                  v0data::PosTrackId, v0data::NegTrackId, // indices of V0 tracks in FullTracks table
+                  v0data::PxPos, v0data::PyPos, v0data::PzPos, v0data::PxNeg, v0data::PyNeg, v0data::PzNeg,
+                  v0data::DCAV0Daughters,
+                  v0data::DCAPosToPV, // this is the impact param wrt prim vtx in xy!
+                  v0data::DCANegToPV, // this is the impact param wrt prim vtx in xy!
```
Are these not accessible from the V0 candidate?

chiarazampolli replied:
Yes, but don't we want them in our table so that it suffices for analysis?

vkucera commented on Analysis/Tasks/PWGHF/CMakeLists.txt:
```diff
@@ -25,7 +25,7 @@ o2_add_dpl_workflow(qa-simple
 o2_add_dpl_workflow(hf-track-index-skims-creator
                     SOURCES HFTrackIndexSkimsCreator.cxx
-                    PUBLIC_LINK_LIBRARIES O2::Framework O2::AnalysisDataModel O2::AnalysisCore O2::DetectorsVertexing ROOT::EG
+                    PUBLIC_LINK_LIBRARIES O2::Framework O2::AnalysisDataModel O2::AnalysisCore O2::DetectorsVertexing O2::AnalysisTasksUtils ROOT::EG 
```
Trailing spaces should be removed (ideally by the editor automatically) everywhere.

chiarazampolli replied:
done

vkucera commented on Analysis/Tasks/PWGHF/HFCandidateCreatorCascade.cxx:
```diff
@@ -0,0 +1,304 @@
+#include "ReconstructionDataFormats/DCA.h"
+#include "ReconstructionDataFormats/V0.h"
+#include "AnalysisTasksUtils/UtilsDebugLcK0Sp.h"
+#include "AnalysisCore/HFSelectorCuts.h"
```
`HFSelectorCuts.h` is not needed.

chiarazampolli replied:
done

vkucera commented on Analysis/Tasks/PWGHF/HFCandidateCreatorCascade.cxx:
```diff
@@ -0,0 +1,304 @@
+using namespace o2;
+using namespace o2::framework;
+using namespace o2::aod::hf_cand_prong2;
+using namespace o2::analysis;
```
`analysis` is not needed.

chiarazampolli replied:
done

vkucera commented on Analysis/Tasks/PWGHF/HFCandidateCreatorCascade.cxx:
```diff
@@ -0,0 +1,304 @@
+
+/// Reconstruction of heavy-flavour cascade decay candidates
+struct HFCandidateCreatorCascade {
+
```
Unnecessary blank line right after `struct HFCandidateCreatorCascade {`.

chiarazampolli replied:
I like it, is there a rule that says I cannot have it?

vkucera commented on Analysis/Tasks/PWGHF/HFCandidateCreatorCascade.cxx:
```diff
@@ -0,0 +1,304 @@
+
+  Produces<aod::HfCandCascBase> rowCandidateBase;
+
+  Configurable<double> d_bZ{"d_bZ", 5., "magnetic field"};
+  Configurable<bool> b_propDCA{"b_propDCA", true, "create tracks version propagated to PCA"};
+  Configurable<double> d_maxR{"d_maxR", 200., "reject PCA's above this radius"};
+  Configurable<double> d_maxDZIni{"d_maxDZIni", 4., "reject (if>0) PCA candidate if tracks DZ exceeds threshold"};
+  Configurable<double> d_minParamChange{"d_minParamChange", 1.e-3, "stop iterations if largest change of any X is smaller than this"};
+  Configurable<double> d_minRelChi2Change{"d_minRelChi2Change", 0.9, "stop iterations is chi2/chi2old > this"};
+  Configurable<bool> b_doValPlots{"b_doValPlots", true, "do validation plots"};
+
+  // for debugging
+#ifdef MY_DEBUG
+  Configurable<std::vector<int>> v_labelK0Spos{"v_labelK0Spos", {729, 2866, 4754, 5457, 6891, 7824, 9243, 9810}, "labels of K0S positive daughters, for debug"};
+  Configurable<std::vector<int>> v_labelK0Sneg{"v_labelK0Sneg", {730, 2867, 4755, 5458, 6892, 7825, 9244, 9811}, "labels of K0S negative daughters, for debug"};
+  Configurable<std::vector<int>> v_labelProton{"v_labelProton", {717, 2810, 4393, 5442, 6769, 7793, 9002, 9789}, "labels of protons, for debug"};
+#endif
```
Underscores are not allowed. Please remove them together with the prefixes everywhere in the new code.

chiarazampolli replied:
done

vkucera commented on Analysis/Tasks/PWGHF/HFCandidateCreatorCascade.cxx:
```diff
@@ -0,0 +1,304 @@
+//#define MY_DEBUG
+
+#ifdef MY_DEBUG
+using MyTracks = aod::BigTracksMC;
```
Can you replace `MyTracks` with something more specific, please? :-)

chiarazampolli replied:
Changed to MyBigTracks, similar to what it starts from.

vkucera commented on Analysis/Tasks/PWGHF/HFCandidateCreatorCascade.cxx:
```diff
@@ -0,0 +1,304 @@
+      const auto& negTrack = v0.negTrack_as<MyTracks>();
+
+#ifdef MY_DEBUG
+      auto protonLabel = bach.mcParticleId();
+      auto labelPos = posTrack.mcParticleId();
+      auto labelNeg = negTrack.mcParticleId();
+      bool isLc = isLcK0SpFunc(protonLabel, labelPos, labelNeg, v_labelProton, v_labelK0Spos, v_labelK0Sneg);
```
The names are confusing here.
* `protonLabel` should be `labelBach`
* These are not labels but indices.

chiarazampolli replied:
Why not "proton"? This is a proton. 
Can you tell me what the difference is? In the MC table, isn't the label the index?

vkucera commented on Analysis/Tasks/PWGHF/HFCandidateCreatorCascade.cxx:
```diff
@@ -0,0 +1,304 @@
+      const auto& bach = casc.index0_as<MyTracks>();
+      auto trackParCovBach = getTrackParCov(bach);
+      const auto& v0 = casc.indexV0_as<o2::aod::V0Datas>();
+      const auto& posTrack = v0.posTrack_as<MyTracks>();
+      const auto& negTrack = v0.negTrack_as<MyTracks>();
```
For consistency, `Pos` and `Neg` should always appear at the end of the names of variables.

chiarazampolli replied:
Why? I mean, one says "positive track", not "track positive"

chiarazampolli replied:
And in addition, in the V0 table it is called "posTrack".

vkucera commented on Analysis/Tasks/PWGHF/HFCandidateCreatorCascade.cxx:
```diff
@@ -0,0 +1,304 @@
+
+    // loop over pairs of track indeces
+    for (const auto& casc : rowsTrackIndexCasc) {
+      const auto& bach = casc.index0_as<MyTracks>();
```
This is a track too. Should be `trackBach`.

chiarazampolli replied:
I don't think it makes a difference, but ok.

vkucera commented on Analysis/Tasks/PWGHF/HFCandidateCreatorCascade.cxx:
```diff
@@ -0,0 +1,304 @@
+    // loop over pairs of track indeces
+    for (const auto& casc : rowsTrackIndexCasc) {
+      const auto& bach = casc.index0_as<MyTracks>();
+      auto trackParCovBach = getTrackParCov(bach);
```
`trackParCovBach` should appear together with the other `trackParCov...`.

chiarazampolli replied:
done, was forgotten

vkucera commented on Analysis/Tasks/PWGHF/HFCandidateCreatorCascade.cxx:
```diff
@@ -0,0 +1,304 @@
+
+      auto posTrackParCov = getTrackParCov(posTrack); // check that MyTracks does not need TracksExtended!
+      auto negTrackParCov = getTrackParCov(negTrack); // check that MyTracks does not need TracksExtended!
+      posTrackParCov.propagateTo(v0.posX(), d_bZ);    // propagate the track to the X closest to the V0 vertex
+      negTrackParCov.propagateTo(v0.negX(), d_bZ);    // propagate the track to the X closest to the V0 vertex
```
Should be `trackParCovPos`, `trackParCovNeg`

chiarazampolli replied:
Where can I find these rules? Just so that next time I do it right from the beginning. For me it is natural to put "pos"/"neg" in front.

chiarazampolli replied:
For now I don't change it, since I would not change the "posTrack" above (mainly because it is like this in the V0 table)

vkucera commented on Analysis/Tasks/PWGHF/HFCandidateCreatorCascade.cxx:
```diff
@@ -0,0 +1,304 @@
+      // fill histograms
+      if (b_doValPlots) {
+        // calculate invariant masses
+        auto arrayMomenta = array{pVecBach, pVecV0};
+        mass2K0sP = RecoDecay::M(arrayMomenta, array{massP, massK0s});
```
`arrayMomenta` is not necessary

chiarazampolli replied:
Done

vkucera commented on Analysis/Tasks/PWGHF/HFCandidateCreatorCascade.cxx:
```diff
@@ -0,0 +1,304 @@
+  Produces<aod::HfCandCascadeMCGen> rowMCMatchGen;
+
+#ifdef MY_DEBUG
+  Configurable<std::vector<int>> v_labelK0Spos{"v_labelK0Spos", {729, 2866, 4754, 5457, 6891, 7824, 9243, 9810}, "labels of K0S positive daughters, for debug"};
+  Configurable<std::vector<int>> v_labelK0Sneg{"v_labelK0Sneg", {730, 2867, 4755, 5458, 6892, 7825, 9244, 9811}, "labels of K0S negative daughters, for debug"};
+  Configurable<std::vector<int>> v_labelProton{"v_labelProton", {717, 2810, 4393, 5442, 6769, 7793, 9002, 9789}, "labels of protons, for debug"};
```
Please remove prefixes and underscores.

vkucera commented on Analysis/Tasks/PWGHF/HFCandidateCreatorCascade.cxx:
```diff
@@ -0,0 +1,304 @@
+
+    // Match reconstructed candidates.
+    for (auto& candidate : candidates) {
+      auto arrayDaughtersV0 = array{candidate.posTrack_as<aod::BigTracksMC>(), candidate.negTrack_as<aod::BigTracksMC>()};
+      auto arrayDaughtersLc = array{candidate.index0_as<aod::BigTracksMC>(), candidate.posTrack_as<aod::BigTracksMC>(), candidate.negTrack_as<aod::BigTracksMC>()};
```
`posTrack_as` and `negTrack_as` should not be called twice. Save the track rows in variables.

vkucera replied:
Still there

chiarazampolli replied:
I hope I fixed them all now.

vkucera commented on Analysis/Tasks/PWGHF/HFCandidateCreatorCascade.cxx:
```diff
@@ -0,0 +1,304 @@
+      LOG(DEBUG) << "\n";
+      LOG(DEBUG) << "Checking MC for candidate!";
+      LOG(DEBUG) << "Looking for K0s";
+      auto labelPos = candidate.posTrack_as<aod::BigTracksMC>().mcParticleId();
+      auto labelNeg = candidate.negTrack_as<aod::BigTracksMC>().mcParticleId();
```
Avoid calling `posTrack_as` again.

vkucera commented on Analysis/Tasks/PWGHF/HFCandidateCreatorCascade.cxx:
```diff
@@ -0,0 +1,304 @@
+      LOG(DEBUG) << "\n";
+      LOG(DEBUG) << "Checking MC for candidate!";
+      LOG(DEBUG) << "Looking for K0s";
+      auto labelPos = candidate.posTrack_as<aod::BigTracksMC>().mcParticleId();
+      auto labelNeg = candidate.negTrack_as<aod::BigTracksMC>().mcParticleId();
+      auto protonLabel = candidate.index0_as<aod::BigTracksMC>().mcParticleId();
```
* Avoid getting the rows again.
* These are particle indices, not labels.
* Call the variables consistently.

chiarazampolli replied:
For now I left "label", I would like to clarify what difference it is label and index. The rest, I changed (already following the previous comment in fact)

chiarazampolli replied:
I hope I fixed everything here...

vkucera commented on Analysis/Tasks/PWGHF/HFCandidateCreatorCascade.cxx:
```diff
@@ -0,0 +1,304 @@
+        MY_DEBUG_MSG(sign, LOG(INFO) << "Lc found with sign " << sign; printf("\n"));
+      }
+
+      rowMCMatchRec(sign);
```
I don't see where you reset `sign = 0`.

vkucera commented on Analysis/Tasks/PWGHF/HFCandidateCreatorCascade.cxx:
```diff
@@ -0,0 +1,304 @@
+        MY_DEBUG_MSG(sign, LOG(INFO) << "Lc in K0S p");
+        arrDaughLcIndex.clear();
+        // checking that the final daughters (decay depth = 3) are p, pi+, pi-
+        RecoDecay::getDaughters(particlesMC, particle, &arrDaughLcIndex, arrDaughLcPDGRef, 3); // best would be to check the K0S daughters
```
You can get `arrDaughLcIndex` directly from `isMatchedMCGen`.

chiarazampolli replied:
I am not sure here: will it then check that it is really Lc --> K0 + p --> K0S + p --> pi+ pi- p?

chiarazampolli replied:
Leaving as it is for now, do you agree?

vkucera commented on Analysis/Tasks/PWGHF/HFLcK0sPCandidateSelector.cxx:
```diff
@@ -0,0 +1,400 @@
+
+  Produces<aod::HFSelLcK0sPCandidate> hfSelLcK0sPCandidate;
+
+  Configurable<double> d_pTCandMin{"d_pTCandMin", 0., "Lower bound of candidate pT"};
+  Configurable<double> d_pTCandMax{"d_pTCandMax", 50., "Upper bound of candidate pT"};
+
+  // PID
+  Configurable<double> d_applyPidTPCMinPt{"d_applyPidTPCMinPt", 4., "Lower bound of track pT to apply TPC PID"};
+  Configurable<double> d_pidTPCMinPt{"d_pidTPCMinPt", 0., "Lower bound of track pT for TPC PID"};
+  Configurable<double> d_pidTPCMaxPt{"d_pidTPCMaxPt", 100., "Upper bound of track pT for TPC PID"};
+  Configurable<double> d_pidCombMaxP{"d_pidCombMaxP", 4., "Upper bound of track p to use TOF + TPC Bayes PID"};
+  Configurable<double> d_nSigmaTPC{"d_nSigmaTPC", 3., "Nsigma cut on TPC only"};
+
+  // track quality
+  Configurable<double> d_TPCNClsFindablePIDCut{"d_TPCNClsFindablePIDCut", 50., "Lower bound of TPC findable clusters for good PID"};
+  Configurable<bool> b_requireTPC{"b_requireTPC", true, "Flag to require a positive Number of found clusters in TPC"};
+
+  // for debugging
+#ifdef MY_DEBUG
+  Configurable<std::vector<int>> v_labelK0Spos{"v_labelK0Spos", {729, 2866, 4754, 5457, 6891, 7824, 9243, 9810}, "labels of K0S positive daughters, for debug"};
+  Configurable<std::vector<int>> v_labelK0Sneg{"v_labelK0Sneg", {730, 2867, 4755, 5458, 6892, 7825, 9244, 9811}, "labels of K0S negative daughters, for debug"};
+  Configurable<std::vector<int>> v_labelProton{"v_labelProton", {717, 2810, 4393, 5442, 6769, 7793, 9002, 9789}, "labels of protons, for debug"};
```
Prefixes and underscores

vkucera commented on Analysis/Tasks/PWGHF/HFLcK0sPCandidateSelector.cxx:
```diff
@@ -0,0 +1,400 @@
+  template <typename T>
+  bool daughterSelection(const T& track) // aren't these checks already in the indexskimscreator?
+  {
+    if (track.sign() == 0) {
+      return false;
+    }
```
Useless

vkucera commented on Analysis/Tasks/PWGHF/HFLcK0sPCandidateSelector.cxx:
```diff
@@ -0,0 +1,400 @@
+    return -1;
+  }
+
+  /// Selection on goodness of daughter tracks
+  /// \note should be applied at candidate selection
+  /// \param track is daughter track
+  /// \return true if track is good
+  template <typename T>
+  bool daughterSelection(const T& track) // aren't these checks already in the indexskimscreator?
+  {
+    if (track.sign() == 0) {
+      return false;
+    }
+    if (b_requireTPC.value && track.tpcNClsFound() == 0) {
+      return false; //is it clusters findable or found - need to check
+    }
+    return true;
+  }
```
Useless

vkucera commented on Analysis/Tasks/PWGHF/HFLcK0sPCandidateSelector.cxx:
```diff
@@ -0,0 +1,400 @@
+  template <typename T>
+  bool applyTPCPID(const T& track)
+  {
+    if (TMath::Abs(track.pt()) < d_applyPidTPCMinPt) {
```
Useless `TMath::Abs`

chiarazampolli replied:
I had not seen this comment before. fixed now.

vkucera commented on Analysis/Tasks/PWGHF/HFLcK0sPCandidateSelector.cxx:
```diff
@@ -0,0 +1,400 @@
+  template <typename T>
+  bool validCombPID(const T& track)
+  {
+    if (TMath::Abs(track.pt()) > d_pidCombMaxP) { // is the pt sign used for the charge? If it is always positive, we should remove the abs
```
Useless `TMath::Abs`

vkucera commented on Analysis/Tasks/PWGHF/HFLcK0sPCandidateSelector.cxx:
```diff
@@ -0,0 +1,400 @@
+  /// \param nPDG is the PDG code of the flavour hypothesis
+  /// \note nPDG=211 pion  nPDG=321 kaon
+  /// \return 1 if successful PID match, 0 if successful PID rejection, -1 if no PID info
+  template <typename T>
+  int selectionPID(const T& track)
```
All the PID selection will be replaced by a central PID selector. All this PID-related code can be removed.

chiarazampolli replied:
Ok, but should I remove it now?

vkucera commented on Analysis/Tasks/PWGHF/HFLcK0sPCandidateSelector.cxx:
```diff
@@ -0,0 +1,400 @@
+      }
+      */
+
+      const auto& bach = candidate.index0_as<MyBigTracks>(); //bachelor track
```
Avoid calling `index0_as` multiple times.

chiarazampolli replied:
done

vkucera commented on Analysis/Tasks/PWGHF/HFTrackIndexSkimsCreator.cxx:
```diff
@@ -120,7 +195,14 @@ struct SelectTracks {
         if ((status_prong & (1 << 1)) && (abs(dca[0]) < dcatoprimxymin_3prong_ptdep || abs(dca[0]) > dcatoprimxymax_3prong)) {
           status_prong = status_prong & ~(1 << 1);
         }
+        if ((status_prong & (1 << 2)) && (abs(dca[0]) < d_dcaToPrimXYMinBach_ptDep || abs(dca[0]) > d_dcaToPrimXYMaxBach)) {
+          MY_DEBUG_MSG(isProtonFromLc,
+                       LOG(INFO) << "proton " << protonLabel << " did not pass DCA cut";
+                       LOG(INFO) << "dca[0] = " << dca[0] << " (lower cut " << d_dcaToPrimXYMinBach_ptDep << ", upper cut " << d_dcaToPrimXYMaxBach << ")";);
+          status_prong = status_prong & ~(1 << 2);
+        }
```
This part needs to be reworked with the array cuts.

chiarazampolli replied:
Yes, for later

vkucera commented on Analysis/Tasks/PWGHF/HFTrackIndexSkimsCreator.cxx:
```diff
@@ -873,9 +963,285 @@ struct HFTrackIndexSkimsCreator {
+          MY_DEBUG_MSG(isK0SfromLc, LOG(INFO) << "K0S with daughters " << labelPos << " and " << labelNeg << ": rejected due to minPt --> pos " << posTrack.pt() << ", neg " << negTrack.pt() << " (cut " << d_ptMin << ")");
+          continue;
+        }
+        if (abs(posTrack.eta()) > d_etaMax || // to the filters? I can't for now, it is not in the tables
+            abs(negTrack.eta()) > d_etaMax) {
```
`std::abs`

chiarazampolli replied:
Everywhere in this class only "abs" is used.

vkucera commented on Analysis/Tasks/PWGHF/HFTrackIndexSkimsCreator.cxx:
```diff
@@ -873,9 +963,285 @@ struct HFTrackIndexSkimsCreator {
+        if (nCand2 == 0) {
+          continue;
+        }
+        std::array<float, 3> pVecCandCasc = {0., 0., 0.};
+        fitter.propagateTracksToVertex();        // propagate the bach and V0 to the Lc vertex
+        fitter.getTrack(0).getPxPyPzGlo(pVecV0); // take the momentum at the Lc vertex
+        fitter.getTrack(1).getPxPyPzGlo(pVecBach);
+
+        pVecCandCasc = array{pVecBach[0] + pVecV0[0],
+                             pVecBach[1] + pVecV0[1],
+                             pVecBach[2] + pVecV0[2]};
+
+        // cascade candidate pT cut
+        if (RecoDecay::Pt(pVecCandCasc) < d_cutCascPtCandMin) {
```
I think I commented this in the previous version. This is way more complicated than it should be.

vkucera commented on Analysis/Tasks/PWGHF/HFTrackIndexSkimsCreator.cxx:
```diff
@@ -873,9 +963,285 @@ struct HFTrackIndexSkimsCreator {
+
+        // invariant mass
+        // calculate invariant masses
+        auto arrMom = array{pVecBach, pVecV0};
+        mass2K0sP = RecoDecay::M(arrMom, array{massP, massK0s});
```
I already commented previously. Why don't you apply the pT and mass cuts before doing the CPU-expensive `fitter.process`?

chiarazampolli replied:
Now changed (I thought I commented already, I don't know why it is still here... ).

vkucera commented on Analysis/Tasks/PWGHF/taskLcK0sP.cxx:
```diff
@@ -0,0 +1,155 @@
+// or submit itself to any jurisdiction.
+
+/// \file taskLcK0sp.cxx
+/// \brief LcK0sp analysis task
```
Λc → K0S p

vkucera commented on Analysis/Tasks/PWGHF/taskLcK0sP.cxx:
```diff
@@ -0,0 +1,155 @@
+struct TaskLcK0sP {
+  HistogramRegistry registry{
+    "registry",
+    {{"hmass", "cascade candidates;inv. mass (#pi K) (GeV/#it{c}^{2});entries", {HistType::kTH1F, {{500, 0.0f, 5.0f}}}},
+     {"hptcand", "cascade candidates;candidate #it{p}_{T} (GeV/#it{c});entries", {HistType::kTH1F, {{100, 0.0f, 10.0f}}}},
+     {"hptbach", "cascade candidates;bachelor #it{p}_{T} (GeV/#it{c});entries", {HistType::kTH1F, {{100, 0.0f, 10.0f}}}},
+     {"hptv0", "cascade candidates;v0 #it{p}_{T} (GeV/#it{c});entries", {HistType::kTH1F, {{100, 0.0f, 10.0f}}}},
+     {"hd0bach", "cascade candidates;bachelor DCAxy to prim. vertex (cm);entries", {HistType::kTH1F, {{100, -1.0f, 1.0f}}}},
+     {"hd0v0pos", "cascade candidates;pos daugh v0 DCAxy to prim. vertex (cm);entries", {HistType::kTH1F, {{100, -5.0f, 5.0f}}}},
+     {"hd0v0neg", "cascade candidates;neg daugh v0 DCAxy to prim. vertex (cm);entries", {HistType::kTH1F, {{100, -5.0f, 5.0f}}}},
+     {"hv0CPA", "cascade candidates;v0 cosine of pointing angle;entries", {HistType::kTH1F, {{110, -0.98f, 1.1f}}}},
+     {"hEta", "cascade candidates;candidate #it{#eta};entries", {HistType::kTH1F, {{100, -2.0f, 2.0f}}}},
+     {"hselectionstatus", "cascade candidates;selection status;entries", {HistType::kTH1F, {{5, -0.5f, 4.5f}}}}}};
```
Fix camelCase

chiarazampolli replied:
hd0xxx changed to hd0Xxx, keeping "d0" small, since it is the way the variable is commonly called. I hope it is fine

vkucera commented on Analysis/Tasks/PWGHF/taskLcK0sP.cxx:
```diff
@@ -0,0 +1,155 @@
+     {"hselectionstatus", "cascade candidates;selection status;entries", {HistType::kTH1F, {{5, -0.5f, 4.5f}}}}}};
+
+  Configurable<int> d_selectionFlagLcK0sp{"d_selectionFlagLcK0sp", 1, "Selection Flag for LcK0sp"};
+  Configurable<double> cutEtaCandMax{"cutEtaCandMax", -1., "max. cand. pseudorapidity"};
```
Shouldn't it be rapidity?

chiarazampolli replied:
For now I use eta. Is it a problem?

vkucera commented on Analysis/Tasks/PWGHF/taskLcK0sP.cxx:
```diff
@@ -0,0 +1,155 @@
+    "registry",
+    {{"hPtRecSig", "cascade candidates (matched);#it{p}_{T}^{rec.} (GeV/#it{c});entries", {HistType::kTH1F, {{100, 0., 10.}}}},
+     {"hPtRecBg", "cascade candidates (unmatched);#it{p}_{T}^{rec.} (GeV/#it{c});entries", {HistType::kTH1F, {{100, 0., 10.}}}},
+     {"hPtGen", "cascade (matched);#it{p}_{T}^{gen.} (GeV/#it{c});entries", {HistType::kTH1F, {{100, 0., 10.}}}},
```
MC particles

chiarazampolli committed:
more comments by Vit

clang-format

chiarazampolli committed:
Further comments implemented

clang-format

leftovers fixed

Clang-format

Few more

Clang-format

removing useless namespace declaration

Optimization

more

clang-format

typo

Remove commented out code

vkucera commented on Analysis/Tasks/PWGLF/CMakeLists.txt:
```diff
@@ -59,8 +59,8 @@ o2_add_dpl_workflow(nuclei-spectra
 o2_add_dpl_workflow(lambdakzerobuilder
-                    SOURCES lambdakzerobuilder.cxx
-                    PUBLIC_LINK_LIBRARIES O2::Framework O2::DetectorsBase O2::AnalysisDataModel O2::AnalysisCore O2::DetectorsVertexing
+                    SOURCES lambdakzerobuilder
+                    PUBLIC_LINK_LIBRARIES O2::Framework O2::DetectorsBase O2::AnalysisDataModel O2::AnalysisCore O2::DetectorsVertexing O2::AnalysisTasksUtils
```
Missing suffix.

chiarazampolli commented:
Hello @ddobrigk , 

This PR touches also your code in PWGLF, where I added some functionalities to do debugging. Could you please take a look and let me know in case you spot anything wrong or that you don't like?

@ginnocen , @iarsene , @jgrosseo : please, do not approve before David's agree that all is fine! (in case the tests are fine, of course).

Thanks in advance!

Chiara

ktf commented on Analysis/Tasks/Utils/include/AnalysisTasksUtils/UtilsDebugLcK0Sp.h:
```diff
@@ -0,0 +1,90 @@
+  }
+
+  /*
+  auto nElements = nPositiveDau;
```
For next time, do not comment code please.

vkucera replied:
Is it the use of `/*` `*/` that is problematic?

chiarazampolli replied:
Removed, but I would also like to know what is wrong/bad with it. 

Chiara

ddobrigk commented:
I took a look at the changes to the v0 tasks - everything looks fine, as only a debug option was added, essentially. Please feel free to carry on :-D

chiarazampolli commented:
Ciao @ddobrigk , 

Thanks! @ginnocen , @iarsene , @jgrosseo , could you merge now? I think the PR got all needed approvals.

Chiara
